### PR TITLE
Emulate Gemalto reader in unit tests

### PR DESCRIPTION
**Pull request by emaxx-google**
Implement fake USB simulation of an arbitrary smart card reader (we took
Gemalto), and write a test that verifies PC/SC-Lite successfully
initializes it and returns via SCardListReaders.

### CODE REVIEW

**Comment by emaxx-google** [comment]
Apologies for a big PR - I didn't see a reasonable way for splitting it up.

File: smart_card_connector_app/src/testing_smart_card_simulation.h [hunk 1]
```diff
@@ -15,16 +15,36 @@
...
 #include <google_smart_card_common/value.h>
 
+#include "third_party/libusb/webport/src/libusb_js_proxy_data_model.h"
+
 namespace google_smart_card {
 
 // Implements fake smart card reader USB devices.
 class TestingSmartCardSimulation final {
  public:
+  // Fake device to semulate.
```
Comment (emaxx-google) [inline smart_card_connector_app/src/testing_smart_card_simulation.h:38]: self-nit: "simulate"

File: smart_card_connector_app/src/testing_smart_card_simulation.cc [hunk 2]
```diff
@@ -77,10 +322,322 @@ void TestingSmartCardSimulation::PostFakeJsReply(RequestId request_id,
...
+    int64_t device_handle,
+    RequestId request_id) {
+  bool device_found = false;
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    DeviceState* device_state =
+        FindDeviceStateByIdAndHandleLocked(device_id, device_handle);
+    if (device_state) {
+      device_found = true;
+      device_state->opened_device_handle = {};
```
Comment (emaxx-google) [inline smart_card_connector_app/src/testing_smart_card_simulation.cc:416]: self-nit: Probably want to assert `device_state->opened_device_handle` is non-null.

File: smart_card_connector_app/src/application_unittest.cc [hunk 3]
```diff
@@ -79,19 +75,82 @@ class SmartCardConnectorApplicationTest : public ::testing::Test {
...
+    smart_card_simulation_.SetDevices(devices);
+  }
+
+  // Returns all "reader_init_add" messages that were sent to JS (they are sent
+  // when a new reader starts to be initialized).
+  const std::vector<std::string>& reader_init_add_messages() const {
+    return reader_init_add_messages_;
+  }
+  // Returns all "reader_finish_add" messages that were sent to JS (they are
+  // sent when a reader initialization finishes).
```
Comment (Fabian-Sommer) [inline smart_card_connector_app/src/application_unittest.cc:90]: optional nit: I think the comments would be better at the variables themselves, not at their getters

File: smart_card_connector_app/src/application_unittest.cc [hunk 4]
```diff
@@ -79,19 +75,82 @@ class SmartCardConnectorApplicationTest : public ::testing::Test {
...
+  StartApplication();
+}
 
 // Test a PC/SC-Lite context can be established and freed, via direct C function
 // calls `SCardEstablishContext()` and `SCardReleaseContext()`.
 // This is an extended version of the "smoke test" as it verifies the daemon
 // successfully started and replies to calls sent over (fake) sockets.
 TEST_F(SmartCardConnectorApplicationTest, InternalApiContextEstablishing) {
+  // Arrange:
+  StartApplication();
```
Comment (Fabian-Sommer) [inline smart_card_connector_app/src/application_unittest.cc:150]: nit: missing an empty line between comment and function call for consistency (or remove the empty lines on the other Act/Arranges, I don't think they are necessary)

File: smart_card_connector_app/src/testing_smart_card_simulation.h [hunk 5]
```diff
@@ -67,52 +72,62 @@ class TestingSmartCardSimulation final {
...
-                                 RequestId request_id);
-
-  optional<LibusbJsTransferResult> HandleOutputBulkTransferLocked(
-      const std::vector<uint8_t>& data_to_send,
-      DeviceState& device_state);
-  optional<LibusbJsTransferResult> HandleInputBulkTransferLocked(
-      int64_t length_to_receive,
-      DeviceState& device_state);
+  // Helper class that provides thread-safe operations with reader states.
+  class Core final {
```
Comment (Fabian-Sommer) [inline smart_card_connector_app/src/testing_smart_card_simulation.h:76]: I think `Core` is a bad name - it doesn't really tell me anything about the purpose of this class. Maybe something like `DeviceStateHelper`? (Note that I have not yet read any code of the class as of now, so I'm just guessing)

**Review by Fabian-Sommer — APPROVED** [review]

File: smart_card_connector_app/src/testing_smart_card_simulation.h (hunk 5 above)
Comment (emaxx-google) [reply smart_card_connector_app/src/testing_smart_card_simulation.h:76]: I see. Maybe `ThreadSafeHandler`?

Comment (Fabian-Sommer) [reply smart_card_connector_app/src/testing_smart_card_simulation.h:76]: Sounds good

**Comment by emaxx-google** [comment]
> (Note that I have not yet read any code of the class as of now, so I'm just guessing)

So do you want to take another look, or the LGTM is up-to-date?

**Comment by Fabian-Sommer** [comment]
LGTM!

**Pull request merged**